src/rgw/rgw_aws.cc

```diff
@@ -0,0 +1,259 @@
+        }
+      }
+    }
+
+    // used in the dtor for message cleanup
+    static void delete_message(const message_t *message) {
+      delete message;
+    }
+
+  public:
+    Manager(size_t _max_clients, size_t max_queue_size,unsigned idle_time_ms, CephContext *_cct) : messageQueue(max_queue_size),
+                                                                                                   clientList(_max_clients),
+                                                                                                   runner(&Manager::run, this),
+                                                                                                   stopped(false), cct(_cct),
+                                                                                                   idle_time(std::chrono::milliseconds(idle_time_ms)){}
+    // dtor wait for thread to stop
+    // then connection are cleaned-up
+    ~Manager() {
+      stopped = true;
+      runner.join();
+      messageQueue.consume_all(delete_message);
+    }
+
+
+    AwsClient connect(const std::string &accessKey,
+                      const std::string &accessSecret,
+                      const std::string &arn,
+                      const std::string &caPath,
+                      bool useSSL) {
+      if (stopped) {
+        return AwsClient();
+      }
+      const client_id_t id(accessKey, accessSecret, arn, useSSL);
+      std::lock_guard lock(clientsLock);
+      const auto it = clientList.find(id);
+      if (it != clientList.end()) {
+        return it->second;
+      }
+
+      // connection not found, creating a new one
+      Aws::Client::ClientConfiguration configuration;
+      if (useSSL) {
+        configuration.scheme = Aws::Http::Scheme::HTTPS;
+        configuration.verifySSL = true;
+      } else {
+        configuration.scheme = Aws::Http::Scheme::HTTP;
+        configuration.verifySSL = false;
+      }
+      Aws::Utils::ARN arn1((Aws::String(arn)));
+      configuration.region = Aws::String(arn1.GetRegion());
+      configuration.maxConnections = 25;
+      configuration.requestTimeoutMs = 3000;
+      configuration.connectTimeoutMs = 3000;
+      configuration.caPath = Aws::String(caPath);
+      Aws::Auth::AWSCredentials credentials;
+      credentials.SetAWSAccessKeyId(Aws::String(accessKey));
+      credentials.SetAWSSecretKey(Aws::String(accessSecret));
+      AwsClient client;
+      if (arn1.GetService() == "lambda") {
+        client = new Aws::Lambda::LambdaClient(credentials, configuration);
+      } else if (arn1.GetService() == "sns") {
+        client = new Aws::SNS::SNSClient(credentials, configuration);
+      }
+      return clientList.emplace(id, client).first->second;
+    }
+
+    int publish(AwsClient client,
+                const std::string &payload,
+                const std::string &resourceARN) {
+      if (stopped) {
+        return STATUS_MANAGER_STOPPED;
+      }
+      if (client == AwsClient()) {
```

Comment:
please create a "static const NO_CLIENT" so it is clear what `AwsClient()` mean

Reply:
how about a typedef?

Reply:
always prefer a `constexpr` variable (or `static const` if `constexpr` is not possible) over a `#define`.